node/src/components/block_proposer.rs

```diff
@@ -380,33 +386,72 @@ impl BlockProposerReady {
     ) -> ProtoBlock {
         let max_transfers = deploy_config.block_max_transfer_count as usize;
         let max_deploys = deploy_config.block_max_deploy_count as usize;
+        let max_block_size_bytes = deploy_config.max_block_size as usize;
+        let block_gas_limit = Gas::from(deploy_config.block_gas_limit);
 
         let mut transfers = Vec::new();
         let mut wasm_deploys = Vec::new();
+        let mut block_gas_running_total = Gas::zero();
+        let mut block_size_running_total = 0usize;
 
         for (hash, deploy_type) in self.sets.pending.iter() {
-            let under_max_transfers = transfers.len() < max_transfers;
-            let under_max_deploys = wasm_deploys.len() < max_deploys;
-            if !under_max_deploys && !under_max_transfers {
+            let at_max_transfers = transfers.len() == max_transfers;
+            let at_max_deploys = wasm_deploys.len() == max_deploys;
+
+            // Early exit if block limits are met.
+            // TODO: break early if gas total is met, but only once we have a constant for the cost
+            // of wasm-less transfers. if block_gas_running_total >= block_gas_limit
+            if block_size_running_total + DEPLOY_APPROX_MIN_SIZE >= max_block_size_bytes
+                || (at_max_deploys && at_max_transfers)
+            {
                 break;
             }
 
-            if self.is_deploy_valid(
-                deploy_type.header(),
+            // Skip this deploy, searching instead for another that may fit the bill.
+            if block_size_running_total + deploy_type.size() > max_block_size_bytes {
+                continue;
+            }
+
+            if !self.is_deploy_valid(
+                &deploy_type.header(),
                 block_timestamp,
                 &deploy_config,
                 &past_deploys,
-            ) && !past_deploys.contains(hash)
-                && !self.sets.finalized_deploys.contains_key(hash)
+            ) || past_deploys.contains(hash)
+                || self.sets.finalized_deploys.contains_key(hash)
```

Comment:
The `continue` criterion above could be added to these three, too.

Reply:
Sorry I don't understand what you mean here.